functions/main.py

```diff
@@ -0,0 +1,150 @@
+        for name in document_data["files"]["arrayValue"]["values"]
+    ]
+    logging.info("Reading source files", extra={"names": names})
+    blobs = list(filter(None, [storage.bucket().get_blob(name) for name in names]))
+
+    size = 0
+    output = {}
+    for blob in blobs:
+        if blob is None:
+            continue
+        size += blob.size or 0
+        if (size / 1e6) > MAX_SIZE:
+            break
+        output[blob.name] = blob.download_as_text()
+    return output
+
+
+def get_int_utcnow() -> int:
+    return int(datetime.utcnow().timestamp())
+
+
+def create_tree_v2(event, context):
+    """Handles new created documents in firestore with path `trees/{treeId}`"""
+    tree_id = context.resource.split("/").pop()
+
+    logging.info(f"Handling new created tree {event} {context} {tree_id}")
+
+    client = firestore.client()
+    document_reference = client.collection("trees").document(tree_id)
+    document_reference.update({"startedDate": get_int_utcnow()})
+
+    try:
+        logging.info("Tree process started")
+        contents = get_contents(event["value"]["fields"])
+        tos = tree_from_strings(list(contents.values()))
+        result = convert_tos_to_json(tos)
+        document_reference.update(
+            {
+                "version": "2",
+                "result": result,
+                "error": None,
+                "finishedDate": get_int_utcnow(),
+            }
+        )
+        logging.info("Tree process finished")
+    except Exception as error:
+        logging.exception("Tree process failed")
+        document_reference.update(
+            {
+                "version": "2",
+                "result": None,
+                "error": str(error),
+                "finishedDate": get_int_utcnow(),
+            }
+        )
+
+
+@on_document_created(document="users/{userId}/trees/{treeId}")
+def process_user_tree(event: Event[DocumentSnapshot | None]) -> None:
+    if event.data is None or (data := event.data.to_dict()) is None:
+        return
+    logging.info(f"handling new created tree {event} {event.data.to_dict()}")
+
+    client = firestore.client()
+    ref = client.document(event.document)
+    ref.update({"startedDate": get_int_utcnow()})
+
+    try:
+        logging.info("Tree process started")
+        contents = get_contents(data)
+        tos = tree_from_strings(list(contents.values()))
+        result = convert_tos_to_json(tos)
+        ref.update(
+            {
+                "version": "2",
+                "result": result,
+                "error": None,
+                "finishedDate": get_int_utcnow(),
+            }
+        )
+        logging.info("Tree process finished")
+    except Exception as error:
+        logging.exception("Tree process failed")
+        ref.update(
+            {
+                "version": "2",
+                "result": None,
+                "error": str(error),
+                "finishedDate": get_int_utcnow(),
```

Comment:
I think that storing total time in milliseconds could be helpful when having the need of researching it through firestore, also using more consistent names for the fields sounds like a good improvement.
```suggestion
    started_timestamp = get_int_utcnow()
    ref.update({"startedTimestamp": started_timestamp})

    try:
        logging.info("Tree process started")
        contents = get_contents(data)
        tos = tree_from_strings(list(contents.values()))
        result = convert_tos_to_json(tos)
        ref.update(
            {
                "version": "2",
                "result": result,
                "error": None,
                "finishedTimestamp": get_int_utcnow(),
                "totalMs": get_int_utcnow() - started_timestamp,
            }
        )
        logging.info("Tree process finished")
    except Exception as error:
        logging.exception("Tree process failed")
        ref.update(
            {
                "version": "2",
                "result": None,
                "error": str(error),
                "finishedTimestamp": get_int_utcnow(),
                "totalMs": get_int_utcnow() - started_timestamp,
```